api/include/opentelemetry/trace/propagation/http_trace_context.h

```diff
@@ -83,168 +65,79 @@ class HttpTraceContext : public TextMapPropagator<T>
-    std::string hex_string = "00-";
-    for (int i = 0; i < 32; i++)
-    {
-      hex_string.push_back(trace_id[i]);
-    }
-    hex_string.push_back('-');
-    for (int i = 0; i < 16; i++)
+    std::array<nostd::string_view, 4> fields{};
+    if (detail::SplitString(trace_parent, '-', fields.data(), 4) != 4)
     {
-      hex_string.push_back(span_id[i]);
+      return SpanContext::GetInvalid();
     }
-    hex_string.push_back('-');
-    for (int i = 0; i < 2; i++)
-    {
-      hex_string.push_back(trace_flags[i]);
-    }
-    setter(carrier, kTraceParent, hex_string);
-    setter(carrier, kTraceState, span_context.trace_state()->ToHeader());
-  }
 
-  static void InjectImpl(Setter setter, T &carrier, const SpanContext &span_context)
-  {
-    InjectTraceHeaders(span_context, carrier, setter);
-  }
-
-  static bool IsValidHex(nostd::string_view string_view)
-  {
-    for (int i = 0; i < string_view.length(); i++)
-    {
-      if (!(string_view[i] >= '0' && string_view[i] <= '9') &&
-          !(string_view[i] >= 'a' && string_view[i] <= 'f'))
-        return false;
-    }
-    return true;
-  }
+    nostd::string_view version_hex     = fields[0];
+    nostd::string_view trace_id_hex    = fields[1];
+    nostd::string_view span_id_hex     = fields[2];
+    nostd::string_view trace_flags_hex = fields[3];
 
-  static SpanContext ExtractContextFromTraceHeaders(nostd::string_view trace_parent,
-                                                    nostd::string_view trace_state)
-  {
-    if (trace_parent.length() != kHeaderSize || trace_parent[kHeaderElementLengths[0]] != '-' ||
-        trace_parent[kHeaderElementLengths[0] + kHeaderElementLengths[1] + 1] != '-' ||
-        trace_parent[kHeaderElementLengths[0] + kHeaderElementLengths[1] +
-                     kHeaderElementLengths[2] + 2] != '-')
+    if (!detail::IsValidHex(version_hex) || !detail::IsValidHex(trace_id_hex) ||
+        !detail::IsValidHex(span_id_hex) || !detail::IsValidHex(trace_flags_hex))
     {
-      std::cout << "Unparseable trace_parent header. Returning INVALID span context." << std::endl;
-      return SpanContext(false, false);
+      return SpanContext::GetInvalid();
     }
-    nostd::string_view version = trace_parent.substr(0, kHeaderElementLengths[0]);
-    nostd::string_view trace_id =
-        trace_parent.substr(kHeaderElementLengths[0] + 1, kHeaderElementLengths[1]);
-    nostd::string_view span_id = trace_parent.substr(
-        kHeaderElementLengths[0] + kHeaderElementLengths[1] + 2, kHeaderElementLengths[2]);
-    nostd::string_view trace_flags = trace_parent.substr(
-        kHeaderElementLengths[0] + kHeaderElementLengths[1] + kHeaderElementLengths[2] + 3);
 
-    if (version == "ff" || trace_id == "00000000000000000000000000000000" ||
-        span_id == "0000000000000000")
-    {
-      return SpanContext(false, false);
-    }
+    TraceId trace_id       = TraceIdFromHex(trace_id_hex);
+    SpanId span_id         = SpanIdFromHex(span_id_hex);
```

Comment:
I didn't test it, but it seems to me this wouldn't catch too short trace and span ids. Like when parsing a header `00-1-1-00`.

Reply:
Added more checks and additional test cases for w3c propagation.